File: src/main/dotnet/erp/rest/RequestEndpoint.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using AspNetCoreWebApi.Entity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using org.domain.commom;
using org.domain.crud.admin;
using static org.domain.commom.DbUtils;
using static org.domain.crud.admin.RequestFilter;

[Route ("rest/request")]
[ApiController]
public class RequestEndpoint : ControllerBase {
	private readonly DbContext entityManager;

	public RequestEndpoint(CrudContext dbContext) {
		this.entityManager = dbContext;
	}

	[HttpPost("create")]
	public Task<ActionResult> Create([FromBody] AspNetCoreWebApi.Entity.Request obj) {
		Console.WriteLine ("Dedicated implement of route {0}", this.RouteData.ToString());
		// TODO : validar se entity.getState() é um State com antecessor vazio.
		if (obj.Date == null) {
			obj.Date = System.DateTime.Now;
		}

		String serviceName = CaseConvert.UnderscoreToCamel(obj.GetType().Name, false);
		return RequestFilter.ProcessCreate<AspNetCoreWebApi.Entity.Request>(this.User.Identity, this.entityManager, serviceName, obj);
	}

	[HttpPut("update")]
	public Task<ActionResult> Update([FromBody] AspNetCoreWebApi.Entity.Request newObj) {
		Console.WriteLine ("Dedicated implement of route {0}", this.RouteData.ToString ());
		String serviceName = CaseConvert.UnderscoreToCamel(newObj.GetType().Name, false);
        // TODO : validar se entity.getState() é um State com antecessor e precedente validos.
		return RequestFilter.GetObject<AspNetCoreWebApi.Entity.Request>(this.User.Identity, this.Request, this.entityManager, serviceName).ContinueWith<ActionResult>(taskOldObj => {
			return RequestFilter.ProcessUpdate(this.User.Identity, this.Request, this.entityManager, serviceName, newObj).ContinueWith<ActionResult>(taskResponse => {
				ActionResult response = taskResponse.Result;

				if (response is OkResult) {
					RequestState stateOld = RequestProductEndpoint.GetRequestState(this.entityManager, taskOldObj.Result.State);
					RequestState state = RequestProductEndpoint.GetRequestState(this.entityManager, newObj.State);
					List<RequestProduct> list = DbUtils.Find<RequestProduct>(this.entityManager, null, QueryMap.Create().AddNext("company", newObj.Company).AddNext("request", newObj.Id), null, null, null).Result;

					foreach (RequestProduct requestProduct in list) {
					}
				}

				return response;
			}).Result;
		});
	}

	[HttpDelete("delete")]
	public Task<ActionResult> Remove() {
		Console.WriteLine ("Dedicated implement of route {0}", this.RouteData.ToString());
		// TODO : validar se entity.getState() é um State de status iniciais que permite exclusão.
		String serviceName = "request";
		return RequestFilter.GetObject<AspNetCoreWebApi.Entity.Request>(this.User.Identity, this.Request, this.entityManager, serviceName).ContinueWith<ActionResult>(taskOldObj => {
			var obj = taskOldObj.Result;
			return RequestFilter.ProcessDelete<AspNetCoreWebApi.Entity.Request> (this.User.Identity, this.Request, this.entityManager, serviceName).ContinueWith<ActionResult> (taskResponse => {
				ActionResult response = taskResponse.Result;

				if (response is OkResult) {
					RequestState stateOld = RequestProductEndpoint.GetRequestState (this.entityManager, obj.State);
					RequestState state = RequestProductEndpoint.GetRequestState (this.entityManager, obj.State);
					List<RequestProduct> list = DbUtils.Find<RequestProduct> (this.entityManager, null, QueryMap.Create ().AddNext ("company", obj.Company).AddNext ("request", obj.Id), null, null, null).Result;

					foreach (RequestProduct requestProduct in list) {
					}
				}

				return response;
			}).Result;
		});
	}

	[HttpGet("query")]
	public Task<ActionResult> Query() {
		Console.WriteLine ("Dedicated implement of route {0}", this.RouteData.ToString());
		// TODO : verificar as condições para carregar as requests encerradas
		String serviceName = "request";
		return RequestFilter.ProcessQuery<AspNetCoreWebApi.Entity.Request>(this.User.Identity, this.Request, this.entityManager, serviceName).ContinueWith<ActionResult> (taskResponse => {
			return taskResponse.Result;
		});
	}

}

File: src/main/java/org/domain/erp/rest/RequestEndpoint.java
package org.domain.erp.rest;

import java.time.LocalDateTime;
import java.util.List;

import javax.annotation.Resource;
import javax.ejb.EJBContext;
import javax.ejb.Stateless;
import javax.ejb.TransactionManagement;
import javax.ejb.TransactionManagementType;
import javax.inject.Named;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.TypedQuery;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.container.AsyncResponse;
import javax.ws.rs.container.Suspended;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import org.domain.crud.admin.RequestFilter;
import org.domain.erp.entity.Request;
import org.domain.erp.entity.RequestProduct;
import org.domain.erp.entity.RequestState;

/**
 *
 */
@Stateless
@Path("/request")
@TransactionManagement(TransactionManagementType.BEAN)
@Named 
public class RequestEndpoint {
	@Resource
	private EJBContext context;	

	@PersistenceContext(unitName = "primary")
	private EntityManager entityManager;

	@POST
	@Path("/create")
	@Consumes("application/json")
	public void create(@Context SecurityContext securityContext, @Context UriInfo uriInfo, Request obj, @Suspended AsyncResponse ar) {
		// TODO : validar se entity.getState() é um State com antecessor vazio.
		if (obj.getDate() == null) {
			obj.setDate(LocalDateTime.now());
		}

		RequestFilter.processCreate(securityContext.getUserPrincipal(), this.context.getUserTransaction(), this.entityManager, "request", obj)
		.exceptionally(error -> Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build())
		.thenAccept(response -> ar.resume(response));
	}

	@PUT
	@Path("/update")
	@Consumes("application/json")
	public void update(@Context SecurityContext securityContext, @Context UriInfo uriInfo, Request newObj, @Suspended AsyncResponse ar) {
		// TODO : validar se entity.getState() é um State com antecessor e precedente validos.
		RequestFilter.<Request>getObject(securityContext.getUserPrincipal(), uriInfo, this.entityManager, "request")
		.thenAccept(oldObj -> {
			RequestFilter.processUpdate(securityContext.getUserPrincipal(), uriInfo, this.context.getUserTransaction(), this.entityManager, "request", newObj)
			.exceptionally(error -> Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build())
			.thenAccept(response -> {
				ar.resume(response);

				if (response.getStatusInfo() == Response.Status.OK) {
					RequestState stateOld = RequestProductEndpoint.getRequestState(this.entityManager, oldObj.getState());
					RequestState state = RequestProductEndpoint.getRequestState(this.entityManager, newObj.getState());
					TypedQuery<RequestProduct> query = this.entityManager.createQuery("from RequestProduct o where o.company = :company and o.request = :request", RequestProduct.class);
					query.setParameter("company", newObj.getCompany());
					query.setParameter("request", newObj.getId());
					List<RequestProduct> list = query.getResultList();

					for (RequestProduct requestProduct : list) {
					}
				}
			});
		})
		.exceptionally(error -> {
			ar.resume(Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build());
			return null;
		});
	}

	@DELETE
	@Path("/delete")
	public void remove(@Context SecurityContext securityContext, @Context UriInfo uriInfo, @Suspended AsyncResponse ar) {
		// TODO : validar se entity.getState() é um State de status iniciais que permite exclusão.
		RequestFilter.<Request>getObject(securityContext.getUserPrincipal(), uriInfo, this.entityManager, "request")
		.thenAccept(obj -> {
			RequestFilter.processDelete(securityContext.getUserPrincipal(), uriInfo, this.context.getUserTransaction(), this.entityManager, "request")
			.exceptionally(error -> Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build())
			.thenAccept(response -> {
				ar.resume(response);

				if (response.getStatusInfo() == Response.Status.OK) {
					RequestState stateOld = RequestProductEndpoint.getRequestState(this.entityManager, obj.getState());
					TypedQuery<RequestProduct> query = this.entityManager.createQuery("from RequestProduct o where o.company = :company and o.request = :request", RequestProduct.class);
					query.setParameter("company", obj.getCompany());
					query.setParameter("request", obj.getId());
					List<RequestProduct> list = query.getResultList();

					for (RequestProduct requestProduct : list) {
					}
				}
			});
		})
		.exceptionally(error -> {
			ar.resume(Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build());
			return null;
		});
	}

	@GET
	@Path("/query")
	@Produces("application/json")
	public void query(@Context SecurityContext context, @Context UriInfo uriInfo, @Suspended AsyncResponse ar) {
		// TODO : verificar as condições para carregar as requests encerradas
		RequestFilter.processQuery(context.getUserPrincipal(), uriInfo, this.entityManager, "request")
		.exceptionally(error -> Response.status(Response.Status.BAD_REQUEST).entity(error.getMessage()).build())
		.thenAccept(response -> {
			ar.resume(response);
		});
	}

}
